### ISSUE TITLE
GPT-OSS MoE's implementation should use `route_norm=True`

### ISSUE THREAD

**Issue by linyuhongg** [pytorch/torchtitan#2258]
### Bug description

The current implementation of gpt-oss sets `route_norm=False`, meaning softmax is applied over all expert logits instead of only the selected topk. However, in the paper, it is stated that "weight the output of each expert by the softmax of the router projection over only the selected experts." 

So we should set `route_norm=True` to apply softmax only over the selected experts. 

### Versions

The code is latest commit a25dd8f880e0759c1496d07ff0ca6973c6dc8b65

**Comment by tianyu-l** [comment]
Thanks.
- Could you show a detailed reference to the paper?
- Would you be interested in submitting a fix?

**Comment by tianyu-l** [comment]
cc @shuhuayu @wwwjn

**Comment by linyuhongg** [comment]
Thanks for the quick response! I can submit a fix for that. 

This is from the section [2.2 Architecture](https://arxiv.org/pdf/2508.10925)
> Mixture-of-Experts: Each MoE block consists of a fixed number of experts (128 for gpt-oss-
120b and 32 for gpt-oss-20b), as well as a standard linear router projection which maps residual
activations to scores for each expert. For both models, we select the top-4 experts for each token
given by the router, and weight the output of each expert by the softmax of the router projection
over only the selected experts. 

Also refer to their repository https://github.com/openai/gpt-oss/blob/599476783c6f88508dab8577808b5ead5cbee8d2/gpt_oss/torch/model.py#L316

Note: it seems like there might be other inconsistencies, for instance, their gate uses the bias term inside the linear layer.

**Comment by shuhuayu** [comment]
@linyuhongg Thanks for pointing this out. Indeed, we should use `route_norm = True` in `models/gpt_oss/__init__.py` to restrict the `softmax` over selected experts only. For the second issue on `gate` bias, in you fix, maybe we can add an attribute `route_gate_bias` to `MoEArgs` in `models/moe/moe.py`, and set this value as `True` for gptoss models in initializations.

**Comment by linyuhongg** [comment]
I also noticed that the rope implementation is not using yarn (torchtitan does implement it for deepseek_v3), this is fine if someone trains from scratch and doesn't use yarn for long-context capabilities. However, torchtitan is used in post-training with torchforge, should I also include this correction?

**Comment by tianyu-l** [comment]
@linyuhongg it's being fixed in https://github.com/pytorch/torchtitan/pull/2216
could you take a look and see if it solves the issue?

**Comment by linyuhongg** [comment]
LGTM
